### PR TITLE
Include heroes in partial join responses' state

### LINKED ISSUE #12995: Faster joins: send `heroes` to handle nameless rooms

### PR DESCRIPTION
**Pull request by DMRobertson**
Fixes #12995, see also https://github.com/matrix-org/complement/pull/557

In the interests of haste, I have cut a few corners:
- reuse existing logic for determining heroes rather than reimplementing it from scratch, at the cost of there being a race between current_state and prev_state
- minimally fixup the complement test to make it pass, without thoroughly checking that we're correctly computing heroes
- similarly for trial test

### CODE REVIEW

File: synapse/federation/federation_server.py [hunk 1]
```diff
@@ -1521,6 +1525,21 @@ def _get_event_ids_for_partial_state_join(
...
+        # and b) even if it's wrong, joining servers will get the full state eventually.
+        heroes = extract_heroes_from_room_summary(summary, join_event.state_key)
+        for hero in heroes:
+            membership_event_id = prev_state_ids.get((EventTypes.Member, hero))
+            if membership_event_id:
+                state_event_ids.add(membership_event_id)
+
     # TODO: return a few more members:
     #   - those with invites
     #   - those that are kicked? / banned
```
Comment (DMRobertson) [inline synapse/federation/federation_server.py:1545]: Unclear if this TODO still stands?

Comment (reivilibre) [reply synapse/federation/federation_server.py:1545]: I guess it's not related to your work here, though ideally this should be an issue so we can talk about it. I'm not convinced the TODO is a good idea.

Comment (reivilibre) [reply synapse/federation/federation_server.py:1545]: oh, I see, we think this comment might've just been talking about the heroes?
Yeah I'd be tempted to knock the comment out.

File: synapse/storage/databases/main/roommember.py [hunk 2]
```diff
@@ -1517,6 +1517,36 @@ def f(txn: LoggingTransaction) -> None:
...
+        r[0] for r in details.get(Membership.JOIN, empty_ms).members if r[0] != me
+    ]
+    invited_user_ids = [
+        r[0] for r in details.get(Membership.INVITE, empty_ms).members if r[0] != me
+    ]
+    gone_user_ids = [
+        r[0] for r in details.get(Membership.LEAVE, empty_ms).members if r[0] != me
+    ] + [r[0] for r in details.get(Membership.BAN, empty_ms).members if r[0] != me]
+
+    # FIXME: order by stream ordering rather than as returned by SQL
```
Comment (reivilibre) [inline synapse/storage/databases/main/roommember.py:1543]: well now they're ordered alphabetically...

**Review by reivilibre — APPROVED** [review]
LGTM

File: synapse/storage/databases/main/roommember.py (hunk 2 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/roommember.py:1543]: Yeah. :shrug:

File: synapse/federation/federation_server.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/federation/federation_server.py]: > Yeah I'd be tempted to knock the comment out. 

have done so.

**Pull request merged**